pkg/etcd/etcdproxy.go

```diff
@@ -0,0 +1,74 @@
+package etcd
+
+import (
+	"net/url"
+
+	"github.com/pkg/errors"
+	"github.com/rancher/k3s/pkg/agent/loadbalancer"
+)
+
+type Proxy interface {
+	Update(addresses []string)
+	ETCDURL() string
+	ETCDAddresses() []string
+	ETCDServerURL() string
+}
+
+func NewETCDProxy(enabled bool, dataDir, etcdURL string) (Proxy, error) {
+	e := &etcdproxy{
+		dataDir:        dataDir,
+		initialETCDURL: etcdURL,
+		etcdURL:        etcdURL,
+	}
+
+	if enabled {
+		lb, err := loadbalancer.New(dataDir, loadbalancer.ETCDServerServiceName, etcdURL, 2379)
+		if err != nil {
+			return nil, err
+		}
+		e.etcdLB = lb
+		e.etcdLBURL = lb.LoadBalancerServerURL()
+	}
+
+	u, err := url.Parse(e.initialETCDURL)
+	if err != nil {
+		return nil, errors.Wrapf(err, "failed to parse %s", e.initialETCDURL)
+	}
+	e.fallbackETCDAddress = u.Host
+	e.etcdPort = u.Port()
+
+	return e, nil
+}
+
+type etcdproxy struct {
+	dataDir   string
+	etcdLBURL string
+
+	initialETCDURL      string
+	etcdURL             string
+	etcdPort            string
+	fallbackETCDAddress string
+	etcdAddresses       []string
+	etcdLB              *loadbalancer.LoadBalancer
+}
+
+func (e *etcdproxy) Update(addresses []string) {
+	if e.etcdLB != nil {
```

Comment:
Is it possible for this value to be nil and update be called? If so, what would the expected behavior be in that situation?

Reply:
I dont think there is a possible way for the load balancer to not be initialized and the the program will not exit with an error, however to avoid any panic I added this condition